homeassistant/components/bryant_evolution/coordinator.py

```diff
@@ -0,0 +1,190 @@
+    # Create a client for an arbitrary (but known-valid) zone to read parameters
+    # that are zone-independent.
+    sys_client = await BryantEvolutionLocalClient.get_client(
+        system_id, zone_ids[0], tty
+    )
+    fan_mode = (await sys_client.read_fan_mode()).lower()
+    (hvac_mode, is_hvac_active) = await sys_client.read_hvac_mode()
+    zones: dict[int, _ZoneState] = {}
+    for zone_id in zone_ids:
+        zones[zone_id] = await _read_zone_state(system_id, zone_id, tty)
+    return _SystemState(
+        fan_mode=fan_mode,
+        hvac_mode=hvac_mode,
+        is_hvac_active=is_hvac_active,
+        zones=zones,
+    )
+
+
+async def _read_zone_state(system_id: int, zone_id: int, tty: str) -> _ZoneState:
+    """Read the state of a given (system_id, zone_id) at tty."""
+    client = await BryantEvolutionLocalClient.get_client(system_id, zone_id, tty)
+    curr_temp = await client.read_current_temperature()
+    htsp = await client.read_heating_setpoint()
+    clsp = await client.read_cooling_setpoint()
+    return _ZoneState(curr_temp=curr_temp, htsp=htsp, clsp=clsp)
+
+
+class EvolutionCoordinator(DataUpdateCoordinator[EvolutionState]):
+    """Class to manage fetching data from the Evolution API on a particular tty."""
+
+    # Note: directly accessing the data member of this class is discouraged.
+    # Prefer using the provided read_ functions, which are less error-prone.
+
+    def __init__(
+        self, hass: HomeAssistant, tty: str, system_zones: list[tuple[int, int]]
+    ) -> None:
+        """Initialize coordinator."""
+        super().__init__(
+            hass,
+            logger=_LOGGER,
+            name=f"Bryant Evolution at {tty}",
+            update_interval=timedelta(minutes=1),
+            update_method=lambda: _read_evolution_state(
+                self.system_zone_tuples, self.tty
+            ),
+            always_update=False,
+        )
+        self.system_zone_tuples = system_zones
+        self.tty = tty
+
+    def read_current_temperature(self, sys_id: int, zone_id: int) -> int:
+        """Return the current temperature."""
+        return self.data.systems[sys_id].zones[zone_id].curr_temp
+
+    def read_fan_mode(self, sys_id: int) -> str:
+        """Return the current fan mode."""
+        return self.data.systems[sys_id].fan_mode
+
+    def read_hvac_mode(self, sys_id: int) -> HVACMode:
+        """Return the current HVAC mode."""
+        evolution_mode = self.data.systems[sys_id].hvac_mode
+        mode_enum = {
+            "HEAT": HVACMode.HEAT,
+            "COOL": HVACMode.COOL,
+            "AUTO": HVACMode.HEAT_COOL,
+            "OFF": HVACMode.OFF,
+        }.get(evolution_mode.upper())
+        if mode_enum is None:
+            raise HomeAssistantError(
+                translation_domain=DOMAIN,
+                translation_key="failed_to_parse_hvac_mode",
+                translation_placeholders={"mode": evolution_mode},
+            )
```

Comment:
You're right, but:

1) mypy can't figure that out ("homeassistant/components/bryant_evolution/coordinator.py:141: error: Missing return statement  [return]" without this case)

2) I think it's good to avoid this kind of cross-function reasoning where possible. https://lexi-lambda.github.io/blog/2019/11/05/parse-don-t-validate/ is a great article.

Reply:
If it failed to talk to the device, the coordinator should raise UpdateFailed. But I'm just seeing that the coordinator doesn't handle any connection or parsing errors.